// @stardust-collective/dag4-network/dist/cjs/web3/default-dag-web3-provider.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DefaultDagWeb3Provider = void 0;
const api_1 = require("../api");
const global_dag_network_1 = require("../global-dag-network");
class DefaultDagWeb3Provider {
    constructor(netInfo) {
        this.loadBalancerApi = new api_1.LoadBalancerApi();
        this.blockExplorerApi = new api_1.BlockExplorerApi();
        if (netInfo) {
            this.setNetwork(netInfo);
        }
    }
    config(netInfo) {
        if (netInfo) {
            this.setNetwork(netInfo);
        }
        else {
            return this.connectedNetwork;
        }
    }
    setNetwork(netInfo) {
        if (this.connectedNetwork !== netInfo) {
            this.connectedNetwork = netInfo;
            if (!this.blockExplorerApi) {
                this.loadBalancerApi = new api_1.LoadBalancerApi();
                this.blockExplorerApi = new api_1.BlockExplorerApi();
            }
            this.blockExplorerApi.config().baseUrl(netInfo.beUrl);
            this.loadBalancerApi.config().baseUrl(netInfo.lbUrl);
        }
    }
    getBalance(address) {
        return __awaiter(this, void 0, void 0, function* () {
            const balObj = yield this.getLoadBalancerApi().getAddressBalance(address);
            return balObj.balance;
        });
    }
    getTransactionCount(address) {
        return __awaiter(this, void 0, void 0, function* () {
            const balObj = yield this.getLoadBalancerApi().getAddressLastAcceptedTransactionRef(address);
            return balObj.ordinal;
        });
    }
    getTransactionHistory(address, limit = 100) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.getBlockExplorerApi().getTransactionsByAddress(address, limit);
        });
    }
    getTokenTransactionHistory(address, limit = 100) {
        return __awaiter(this, void 0, void 0, function* () {
            return [];
        });
    }
    getTokenAddressBalances(addresses, tokenContractAddress) {
        return __awaiter(this, void 0, void 0, function* () {
            return {};
        });
    }
    getLoadBalancerApi() {
        return this.loadBalancerApi || global_dag_network_1.globalDagNetwork.loadBalancerApi;
    }
    getBlockExplorerApi() {
        return this.blockExplorerApi || global_dag_network_1.globalDagNetwork.blockExplorerApi;
    }
}
exports.DefaultDagWeb3Provider = DefaultDagWeb3Provider;
