// @stardust-collective/dag4-core/dist/esm/cross-platform/clients/fetch.http.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
export class FetchRestService {
    constructor(httpClient) {
        const defaultFetch = (typeof window !== 'undefined' && window.hasOwnProperty('fetch')) ? window.fetch.bind(window) : undefined;
        this.httpClient = httpClient || defaultFetch;
    }
    invoke(options) {
        return this.makeServiceRequest(this.buildRequest(options));
    }
    buildRequest(options) {
        const paramStr = options.queryParams && this.serialize(options.queryParams);
        if (paramStr) {
            options.url = `${options.url}?${paramStr}`;
        }
        const httpHeaders = {
            Accept: 'application/json',
            'Content-Type': 'application/json',
        };
        if (options.authToken && !options.noAuthHeader) {
            httpHeaders.Authorization = options.authToken;
        }
        if (options.headers) {
            Object.keys(options.headers).forEach((key) => {
                httpHeaders[key] = options.headers[key];
            });
        }
        if (options.body) {
            const contentType = httpHeaders['Content-Type'];
            if (contentType === 'application/x-www-form-urlencoded') {
                options.body = this.serialize(options.body);
            }
            else if (contentType === 'application/json') {
                options.body = JSON.stringify(options.body);
            }
        }
        return {
            url: options.url,
            body: options.body,
            headers: httpHeaders,
            method: options.method,
            transformResponse: options.transformResponse,
        };
    }
    // eslint-disable-next-line class-methods-use-this
    makeServiceRequest(options) {
        return new Promise((resolve, reject) => {
            this.httpClient(options.url, options)
                .then((res) => __awaiter(this, void 0, void 0, function* () {
                if (res.status !== 200) {
                    const text = yield res.text();
                    // let error = new Error(text);
                    //
                    // error.status = res.status;
                    // error.statusText = res.statusText;
                    throw new Error(text);
                }
                return res.text();
            }))
                .then(body => {
                try {
                    body = JSON.parse(body);
                }
                catch (_a) { }
                if (options.transformResponse) {
                    resolve(options.transformResponse(body));
                }
                else {
                    resolve(body);
                }
            })
                .catch(err => {
                reject(err);
            });
        });
    }
    // eslint-disable-next-line class-methods-use-this
    serialize(obj) {
        if (obj) {
            const keyMap = Object.keys(obj).map((key) => {
                return `${encodeURIComponent(key)}=${encodeURIComponent(obj[key])}`;
            });
            return keyMap.join('&');
        }
        return '';
    }
}
